src/xrpld/app/rdb/backend/detail/Node.cpp

```
@@ -254,7 +254,13 @@ saveValidatedLedger(
 
         if (app.config().useTxTables())
         {
-            auto db = txnDB.checkoutDb();
+            if (!txnDB)
+            {
+                JLOG(j.fatal()) << "TxTables db isn't available";
+                Throw<std::runtime_error>("TxTables db isn't available");
```

Now a runtime error is a change in behaviour of the program. I am not sure if this should be the expected behaviour, so I will let others confirm this.

---

Disagree.  Adding check `if (app.config().useTxTables() && txnDB) ` will silently continue the code which is change of the current behavior.  That's why I change it to throwing error, at least it will reproduce debug behavior, cause release behavior is undefined in case of nullptr.

---

Since the old code was buggy, we don't need to consider it's behaviour anymore. You now have the power to decide the new behaviour of the new code. And throwing a runtime error is such change. That's why it is important to consider if we want to crash the app and let it be relaunched, or throw then catch and then retry creating the DB connection at this point. Or even just not save the ledger state here and spit an error message(discard the transaction? Initiate a resync?). And such decisions need to be agreed upon and well documented. Sometimes such trivial looking changes could have an unexpected impact on the product.

---

This additional check `(if (!txnDB)) ` is pure placebo, it will never be reached with the current code (because of exception throwing while establishing connection to the DB). So the behavior is exactly the same as before. If someone will change behavior and will allow empty DB pointer, they will get an exception and will know that they need to make changes to this part of the code too.
In case of  ` if (app.config().useTxTables() && txnDB)`  check  and possible future code change - the code  will silently proceed skipping initialization, and developer will miss that they introduce silent bug.
I don't quite understand what is your point ?

---

My point is, just deciding to throw a runtime error is the easy way out but may not be the right way forward. Hence it is important to understand all potential routes which could lead to here(this part of the code) and then decide on potential way forward.

For now, this might be good enough. But we should definitely have a more detailed discussion about such patterns and fixing rest of the code later.